File: src/darwindeck/evolution/population.py
"""Population management with diversity tracking (Phase 4)."""

from dataclasses import dataclass
from typing import List
import random
from darwindeck.genome.schema import GameGenome


# Diversity threshold for warning
DIVERSITY_THRESHOLD = 0.1


@dataclass
class Individual:
    """Individual in population with genome and fitness."""
    genome: GameGenome
    fitness: float = 0.0
    evaluated: bool = False


def genome_distance(g1: GameGenome, g2: GameGenome) -> float:
    """
    Compute distance between two genomes (0.0 = identical, 1.0 = maximally different).

    Uses Hamming distance on key structural features.

    Args:
        g1: First genome
        g2: Second genome

    Returns:
        Distance in range [0.0, 1.0]
    """
    distance = 0.0
    total_features = 0

    # 1. Turn structure phase count
    phase_diff = abs(len(g1.turn_structure.phases) - len(g2.turn_structure.phases))
    distance += min(1.0, phase_diff / 5.0)  # Normalize by max expected diff
    total_features += 1

    # 2. Special effects count
    effect_diff = abs(len(g1.special_effects) - len(g2.special_effects))
    distance += min(1.0, effect_diff / 3.0)
    total_features += 1

    # 3. Win conditions count
    win_diff = abs(len(g1.win_conditions) - len(g2.win_conditions))
    distance += min(1.0, win_diff / 2.0)
    total_features += 1

    # 4. Max turns (normalized)
    turns_diff = abs(g1.max_turns - g2.max_turns) / 1000.0
    distance += min(1.0, turns_diff)
    total_features += 1

    # 5. Setup differences
    card_diff = abs(g1.setup.cards_per_player - g2.setup.cards_per_player)
    distance += min(1.0, card_diff / 26.0)
    total_features += 1

    return distance / total_features


class Population:
    """Population of game genomes with diversity tracking."""

    def __init__(self, individuals: List[Individual]):
        """Initialize population.

        Args:
            individuals: List of individuals in population
        """
        self.individuals = individuals
        self.generation = 0

    def compute_diversity(self) -> float:
        """
        Compute population diversity metric using pairwise distances.

        Higher = more diverse, Lower = converged

        Returns:
            Diversity score in range [0.0, 1.0]
        """
        if len(self.individuals) < 2:
            return 0.0

        # Compute average pairwise distance
        total_distance = 0.0
        pair_count = 0

        # Sample pairs (all pairs for small populations, random sample for large)
        if len(self.individuals) <= 50:
            # Small population: check all pairs
            for i in range(len(self.individuals)):
                for j in range(i + 1, len(self.individuals)):
                    total_distance += genome_distance(
                        self.individuals[i].genome,
                        self.individuals[j].genome
                    )
                    pair_count += 1
        else:
            # Large population: sample 100 random pairs
            for _ in range(100):
                i, j = random.sample(range(len(self.individuals)), 2)
                total_distance += genome_distance(
                    self.individuals[i].genome,
                    self.individuals[j].genome
                )
                pair_count += 1

        if pair_count == 0:
            return 0.0

        avg_distance = total_distance / pair_count
        return avg_distance  # Already in 0-1 range

    def check_diversity_crisis(self) -> bool:
        """Check if diversity has collapsed.

        Returns:
            True if diversity below threshold (crisis), False otherwise
        """
        diversity = self.compute_diversity()
        return diversity < DIVERSITY_THRESHOLD

    def get_best_individual(self) -> Individual:
        """Get individual with highest fitness.

        Returns:
            Best individual
        """
        return max(self.individuals, key=lambda ind: ind.fitness)

    def get_average_fitness(self) -> float:
        """Get average fitness across population.

        Returns:
            Average fitness
        """
        if not self.individuals:
            return 0.0

        evaluated = [ind for ind in self.individuals if ind.evaluated]
        if not evaluated:
            return 0.0

        return sum(ind.fitness for ind in evaluated) / len(evaluated)

File: src/gosim/evolution/population.go
package evolution

import (
	"math"
	"math/rand"

	"github.com/signalnine/darwindeck/gosim/evolution/fitness"
	"github.com/signalnine/darwindeck/gosim/genome"
)

// DiversityThreshold is the threshold below which diversity is considered critical.
const DiversityThreshold = 0.1

// Individual represents a single genome with its fitness score.
type Individual struct {
	Genome         *genome.GameGenome
	Fitness        float64
	Evaluated      bool
	FitnessMetrics *fitness.FitnessMetrics // Full metrics breakdown
}

// Clone creates a deep copy of the individual.
func (ind *Individual) Clone() *Individual {
	clone := &Individual{
		Genome:    ind.Genome.Clone(),
		Fitness:   ind.Fitness,
		Evaluated: ind.Evaluated,
	}
	if ind.FitnessMetrics != nil {
		metricsCopy := *ind.FitnessMetrics
		clone.FitnessMetrics = &metricsCopy
	}
	return clone
}

// Population represents a collection of individuals.
type Population struct {
	Individuals []*Individual
	Generation  int
}

// NewPopulation creates a new population from a list of individuals.
func NewPopulation(individuals []*Individual) *Population {
	return &Population{
		Individuals: individuals,
		Generation:  0,
	}
}

// Size returns the number of individuals in the population.
func (p *Population) Size() int {
	return len(p.Individuals)
}

// GetBestIndividual returns the individual with the highest fitness.
func (p *Population) GetBestIndividual() *Individual {
	if len(p.Individuals) == 0 {
		return nil
	}

	best := p.Individuals[0]
	for _, ind := range p.Individuals[1:] {
		if ind.Fitness > best.Fitness {
			best = ind
		}
	}
	return best
}

// GetAverageFitness returns the average fitness of evaluated individuals.
func (p *Population) GetAverageFitness() float64 {
	if len(p.Individuals) == 0 {
		return 0.0
	}

	var sum float64
	var count int
	for _, ind := range p.Individuals {
		if ind.Evaluated {
			sum += ind.Fitness
			count++
		}
	}

	if count == 0 {
		return 0.0
	}
	return sum / float64(count)
}

// ComputeDiversity calculates population diversity using pairwise distances.
// Higher = more diverse, Lower = converged.
// Returns diversity score in range [0.0, 1.0].
func (p *Population) ComputeDiversity() float64 {
	if len(p.Individuals) < 2 {
		return 0.0
	}

	var totalDistance float64
	var pairCount int

	if len(p.Individuals) <= 50 {
		// Small population: check all pairs
		for i := 0; i < len(p.Individuals); i++ {
			for j := i + 1; j < len(p.Individuals); j++ {
				totalDistance += GenomeDistance(
					p.Individuals[i].Genome,
					p.Individuals[j].Genome,
				)
				pairCount++
			}
		}
	} else {
		// Large population: sample 100 random pairs
		for k := 0; k < 100; k++ {
			i := rand.Intn(len(p.Individuals))
			j := rand.Intn(len(p.Individuals))
			if i == j {
				j = (i + 1) % len(p.Individuals)
			}
			totalDistance += GenomeDistance(
				p.Individuals[i].Genome,
				p.Individuals[j].Genome,
			)
			pairCount++
		}
	}

	if pairCount == 0 {
		return 0.0
	}

	return totalDistance / float64(pairCount)
}

// CheckDiversityCrisis returns true if diversity has collapsed.
func (p *Population) CheckDiversityCrisis() bool {
	return p.ComputeDiversity() < DiversityThreshold
}

// GetUnevaluated returns all individuals that haven't been evaluated.
func (p *Population) GetUnevaluated() []*Individual {
	var unevaluated []*Individual
	for _, ind := range p.Individuals {
		if !ind.Evaluated {
			unevaluated = append(unevaluated, ind)
		}
	}
	return unevaluated
}

// SortByFitness returns individuals sorted by fitness (descending).
func (p *Population) SortByFitness() []*Individual {
	sorted := make([]*Individual, len(p.Individuals))
	copy(sorted, p.Individuals)

	// Simple insertion sort (stable, works well for partially sorted data)
	for i := 1; i < len(sorted); i++ {
		j := i
		for j > 0 && sorted[j-1].Fitness < sorted[j].Fitness {
			sorted[j-1], sorted[j] = sorted[j], sorted[j-1]
			j--
		}
	}
	return sorted
}

// GenomeDistance computes distance between two genomes (0.0 = identical, 1.0 = maximally different).
// Uses Hamming distance on key structural features.
func GenomeDistance(g1, g2 *genome.GameGenome) float64 {
	var distance float64
	var totalFeatures float64

	// 1. Turn structure phase count
	phaseDiff := math.Abs(float64(len(g1.TurnStructure.Phases) - len(g2.TurnStructure.Phases)))
	distance += math.Min(1.0, phaseDiff/5.0) // Normalize by max expected diff
	totalFeatures++

	// 2. Special effects count
	effectDiff := math.Abs(float64(len(g1.Effects) - len(g2.Effects)))
	distance += math.Min(1.0, effectDiff/3.0)
	totalFeatures++

	// 3. Win conditions count
	winDiff := math.Abs(float64(len(g1.WinConditions) - len(g2.WinConditions)))
	distance += math.Min(1.0, winDiff/2.0)
	totalFeatures++

	// 4. Max turns (normalized)
	turnsDiff := math.Abs(float64(g1.TurnStructure.MaxTurns-g2.TurnStructure.MaxTurns)) / 1000.0
	distance += math.Min(1.0, turnsDiff)
	totalFeatures++

	// 5. Setup differences (cards per player)
	cardDiff := math.Abs(float64(g1.Setup.CardsPerPlayer - g2.Setup.CardsPerPlayer))
	distance += math.Min(1.0, cardDiff/26.0)
	totalFeatures++

	return distance / totalFeatures
}
